[Acroform] Use the full path to find the node in the XFA datasets where to store the value

calixteman opened the pull request:
I noticed several 'Path not found' errors because of a field called #subform[2]. From the XFA specs, the hash is used for a class of elements in the template tree. When we're looking for a node in the datasets tree, it doesn't make sense to search for a class. Hence the path element starting with a hash are just skipped.

calixteman committed:
[Acroform] Use the full path to find the node in the XFA datasets where to store the value

I noticed several 'Path not found' errors because of a field called #subform[2].
From the XFA specs, the hash is used for a class of elements in the template tree.
When we're looking for a node in the datasets tree, it doesn't make sense to search
for a class. Hence the path element starting with a hash are just skipped.

Snuffleupagus commented on src/core/annotation.js:
```diff
@@ -1125,7 +1125,12 @@ class Annotation {
       if (loopDict.has("T")) {
-        fieldName.unshift(stringToPDFString(loopDict.get("T")));
+        const t = stringToPDFString(loopDict.get("T"));
+        if (!t.startsWith("#")) {
```
Does this apply to the `T`-entry at line 1100 above as well, or only here?

calixteman replied:
A final field isn't a container, hence it should have a name.

Snuffleupagus reviewed (commented):
Can you please add a test-case, to avoid this breaking in the future?

calixteman commented:
> Can you please add a test-case, to avoid this breaking in the future?

It's a bit hard to have a test.
We generally don't use the value from the xfa datasets except when it's missing from the acroform.
I thought writing a unit test and check that a string is in the saved data but the pdf is encrypted.
I don't have any tool to generate a file with both xfa and acroform (even none to just generate a xfa).

So I see two solutions to have a test use qpdf to have a clear pdf and just check a xml string is included in the file or try to write a pdf by hand. I'm not super excited by the 2nd solution and I'm not sure if we can attach a pdf to the gh issue where encryption has been removed.
Maybe I'm just overlooking something easy... it's bit late here.

Do you have any ideas ?

Snuffleupagus commented:
How about a unit-test then, just so that the new code isn't completely untested?

Note that we've got a number of unit-tests specifically for the `_constructFieldName`-method, and I'd hope that it's not too complicated to add one that covers the new code; see https://github.com/mozilla/pdf.js/blob/e676c9388df184151eecd81ec9592fd857cfeb73/test/unit/annotation_spec.js#L1308-L1385

calixteman commented:
I managed to write a test but I had to a new function in the api.
At least the test covers exactly what it should.

Snuffleupagus commented on test/unit/api_spec.js:
```diff
@@ -1903,6 +1903,46 @@ describe("api", function () {
+        datasets.node,
+        "xfa:data.PPTC_153.Page1.PersonalInformation.TitleAndNameInformation.PersonalInfo.Surname.#text"
+      );
+      expect(surName.nodeValue).toEqual(value);
```
Please remember to clean-up everything at the end:
```suggestion
      expect(surName.nodeValue).toEqual(value);
      
      await loadingTask.destroy();
```

Snuffleupagus commented on src/display/api.js:
```diff
@@ -1091,6 +1091,17 @@ class PDFDocumentProxy {
+   * For testing purposes.
+   * @returns {Promise<Object> | null} A XML tree for the xfa:datasets.
+   */
+  getXFADatasets() {
+    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
+      return this._transport.getXFADatasets();
+    }
+    return null;
+  }
```
You might be able to limit this method to *only* the builds where it makes sense, by instead adding something like the following at the end of the `PDFDocumentProxy`-constructor:
```javascript
if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
  Object.defineProperty(this, "getXFADatasets", {
    value: () => {
      return this._transport.getXFADatasets();
    },
  });
}
```

Snuffleupagus commented on src/display/api.js:
```diff
@@ -3030,6 +3041,13 @@ class WorkerTransport {
     return this.messageHandler.sendWithPromise("GetMarkInfo", null);
   }
 
+  getXFADatasets() {
+    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
+      return this.messageHandler.sendWithPromise("GetXFADatasets", null);
+    }
+    return null;
+  }
+
```
You might be able to limit this method to *only* the builds where it makes sense, by instead adding something like the following at the end of the `WorkerTransport`-constructor:
```javascript
if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
  Object.defineProperty(this, "getXFADatasets", {
    value: () => {
      return this.messageHandler.sendWithPromise("GetXFADatasets", null);
    },
  });
}
```

calixteman commented:
When I wrote the code to save data in the xfa datasets it was with f1040 form in mind.
So I tested with it and my patch broke saving for it: the paths to the fields are wrong but the final name is enough to identify them.
Hence, I fixed the issue and I added a test for f1040.

Snuffleupagus reviewed (approved):
r=me, thanks for adding tests!

Snuffleupagus commented:
/botio test

calixteman commented:
Strange...
I fixed the md5.

calixteman commented:
/botio test
